### PR TITLE
Notice lancero problem

### PR DESCRIPTION
**Pull request by ggggggggg**
This pull request allows a source to stop due to an internal error, and adds checking for row number, column number and lsync to the lancero source. It may be too strict, but will be easy to modify it that i the case.

It also adds `ErroringSource` to test this behavior.

In the process I found that our tests could fail due to a `PublishSync` failure, and fixed those cases with `defer source.Stop()` in a few places.

### CODE REVIEW

File: data_source.go [hunk 1]
```diff
@@ -477,20 +490,20 @@ func (ds *AnySource) PrepareRun() error {
 
 // Stop ends the data supply.
 func (ds *AnySource) Stop() error {
-	if ds.Running() {
-		close(ds.abortSelf)
+	if !ds.Running() {
+		return fmt.Errorf("Anysource not running, cannot stop")
 	}
-	ds.runDone.Wait()
```
Comment (joefowler) [inline data_source.go]: Why isn't the new code calling `ds.Wait()` ?

File: lancero_source.go [hunk 2]
```diff
@@ -468,13 +475,29 @@ func (ls *LanceroSource) distributeData() {
...
+		if ncols != dev.ncols || nrows != dev.nrows || lsync != dev.lsync || framesUsed <= 0 {
+			fmt.Printf("have ncols %v, nrows %v, lsync %v, framesUsed %v. had ncols %v, nrows %v, lsync %v, blockingReadCount %v\n",
+				ncols, nrows, lsync, framesUsed, dev.ncols, dev.nrows, dev.lsync, ls.blockingReadCount)
+			if ls.blockingReadCount > 1 {
+				// lsync calculation fails on first few reads so don't error
+				return fmt.Errorf("have ncols %v, nrows %v, lsync %v, framesUsed %v. had ncols %v, nrows %v, lsync %v",
+					ncols, nrows, lsync, framesUsed, dev.ncols, dev.nrows, dev.lsync)
+			}
+		}
+	}
```
Comment (joefowler) [inline lancero_source.go]: This block looks expensive. Is it? Did you profile? Also, it looks like it is repeated from somewhere.

Comment (joefowler) [reply lancero_source.go]: Also, is it really a good idea, given that LSYNC is computed from timing between reads, which is susceptible to errors when latency is large?

File: rpc_server.go [hunk 3]
```diff
@@ -261,14 +267,28 @@ func (s *SourceControl) Stop(dummy *string, reply *bool) error {
...
-	s.status.Running = false
-	s.activeSource = nil
+	s.observeSourceStop()
 	*reply = true
 	s.broadcastStatus()
 	*reply = true
 	return nil
 }
 
+// observeSourceStop modifies s to make future messag
```
Comment (joefowler) [inline rpc_server.go]: Is this comment complete? Well, obviously it isn't.

**Comment by joefowler** [comment]
Note also that this fixes #77 (or it seems that way to me).

File: data_source.go (hunk 1 above)
Comment (ggggggggg) [reply data_source.go]: I'm not sure what `ds.Wait()` is, I couldn't find the definition, but the tests panic when I use it.

File: lancero_source.go (hunk 2 above)
Comment (ggggggggg) [reply lancero_source.go]: I have run a lancero source with this block in place and it worked at 200 MB/s. I haven't profiled it, but I don't think it is expensive unless it is going to fail anyway. `FindFrameBits` only scans through two frames if it is succesful. I am a bit worried we will occsionally get lsync wrong due to random timing variation, and in that case I plan to add a `succesiveErrorsCount` or similar.

File: data_source.go (hunk 1 above)
Comment (joefowler) [reply data_source.go]: It's a `sync.WaitGroup`. It lets something wait on multiple events to happen, potentially in separate goroutines. Not sure whether we need it here.

Comment (ggggggggg) [reply data_source.go]: `runeDone` is the `sync.WaitGroup` and is what I'm waiting on here. I think `ds.Wait()` is actually the function I'm defining here, and I was causing a stack overflow to get those panics.

File: rpc_server.go (hunk 3 above)
Comment (ggggggggg) [reply rpc_server.go]: fixed

**Pull request merged**